twilight-gateway/src/stream.rs

```diff
@@ -0,0 +1,146 @@
+//! Streaming utilities for initializing clusters of shards.
+
+use crate::{error::ShardInitializeError, Config, Shard, ShardId};
+use futures_util::stream::{FuturesUnordered, Stream};
+use std::{
+    error::Error,
+    fmt::{Display, Formatter, Result as FmtResult},
+};
+use twilight_http::Client;
+
+/// Failure when fetching the recommended number of shards to use from Discord's
+/// REST API.
+#[derive(Debug)]
+pub struct StartRecommendedError {
+    /// Type of error.
+    pub(crate) kind: StartRecommendedErrorType,
+    /// Source error if available.
+    pub(crate) source: Option<Box<dyn Error + Send + Sync>>,
+}
+
+impl Display for StartRecommendedError {
+    fn fmt(&self, f: &mut Formatter<'_>) -> FmtResult {
+        match self.kind {
+            StartRecommendedErrorType::Deserializing => {
+                f.write_str("payload isn't a recognized type")
+            }
+            StartRecommendedErrorType::Request => f.write_str("request failed to complete"),
+        }
+    }
+}
+
+impl Error for StartRecommendedError {
+    fn source(&self) -> Option<&(dyn Error + 'static)> {
+        self.source
+            .as_ref()
+            .map(|source| &**source as &(dyn Error + 'static))
+    }
+}
+
+/// Type of [`StartRecommendedError`] that occurred.
+#[derive(Debug)]
+pub enum StartRecommendedErrorType {
+    /// Received gateway event failed to be deserialized.
+    ///
+    /// The message payload is likely an unrecognized type that is not yet
+    /// supported.
+    Deserializing,
+    /// Requesting recommended shards from Discord's REST API failed.
+    ///
+    /// May be due to something such as a network or authentication issue.
+    Request,
+}
+
+/// Start a range of shards with provided configuration for each shard.
+///
+/// Lower end of the range must be less than the higher end. The higher end of
+/// the range is exclusive.
+///
+/// # Panics
+///
+/// Panics if the lower end of the range is equal to the higher end of the
+/// range or the total isn't greater than the lower or higher end of the range.
+pub fn start_range(
```

Comment:
Worth publically documenting that this method shares the TLS connection between shards, reducing memory usage